chainstate/src/detail/transaction_verifier/mod.rs

```diff
@@ -489,6 +577,212 @@ impl<'a, S: BlockchainStorageRead> TransactionVerifier<'a, S> {
+
+    // Calc not burned tokens that were placed in OutputValue::TokenTransfer after partial burn
+    fn get_change_amount(tx: &Transaction, burn_token_id: &TokenId) -> Result<Amount, TokensError> {
+        let change_amount = tx
+            .outputs()
+            .iter()
+            .filter_map(|x| match x.value() {
+                OutputValue::Coin(_) => None,
+                OutputValue::Token(asset) => match asset {
+                    TokenData::TokenTransferV1 { token_id, amount } => {
+                        (token_id == burn_token_id).then_some(amount)
+                    }
+                    TokenData::TokenIssuanceV1 {
+                        token_ticker: _,
+                        amount_to_issue: _,
+                        number_of_decimals: _,
+                        metadata_uri: _,
+                    }
+                    | TokenData::TokenBurnV1 {
+                        token_id: _,
+                        amount_to_burn: _,
+                    } => None,
+                },
+            })
+            .try_fold(Amount::from_atoms(0), |accum, output| accum + *output)
+            .ok_or(TokensError::CoinOrTokenOverflow)?;
+        Ok(change_amount)
+    }
+
+    fn register_issuance(&mut self, tx: &Transaction) -> Result<(), TokensError> {
+        let token_id = token_id(tx).ok_or(TokensError::TokenIdCantBeCalculated)?;
+
+        let _tokens_op = match self.tokens_cache.entry(token_id) {
+            Entry::Occupied(entry) => entry.into_mut(),
+            Entry::Vacant(entry) => entry.insert(CachedTokensOperation::Write(tx.get_id())),
+        };
+        Ok(())
+    }
+
+    fn undo_issuance(&mut self, token_id: TokenId) -> Result<(), TokensError> {
+        match self.tokens_cache.entry(token_id) {
+            Entry::Occupied(entry) => {
+                let tokens_op = entry.into_mut();
+                *tokens_op = CachedTokensOperation::Erase;
+            }
+            Entry::Vacant(entry) => {
+                entry.insert(CachedTokensOperation::Erase);
+            }
+        }
+        Ok(())
+    }
+
+    fn check_connected_tokens(
+        &mut self,
+        token_data: &TokenData,
+        tx: &Transaction,
+        block: &Block,
+    ) -> Result<(), TokensError> {
+        match token_data {
+            TokenData::TokenTransferV1 { token_id, amount } => {
+                self.check_connected_transfer_data(block.get_id(), tx, token_id, amount)?;
+            }
+            TokenData::TokenIssuanceV1 {
+                token_ticker,
+                amount_to_issue,
+                number_of_decimals,
+                metadata_uri,
+            } => {
+                self.check_connected_issue_data(
+                    token_ticker,
+                    amount_to_issue,
+                    number_of_decimals,
+                    metadata_uri,
+                    tx.get_id(),
+                    block.get_id(),
+                )?;
+                self.register_issuance(tx)?;
```

Comment:
This function is "check" yet it's mutable and "registers" issuance mutably. Please move issuance registration and separate it from "checking". Mutable functions must be isolated and as small as possible.

Reply:
Now, token data checks in `check_block`. When we connect a transaction, we check transferred amounts of coin and tokens. Function `register_tokens` isolated from checks.